File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for(int i = 0; i < board.size(); i++)
        {
            set<char> s;
            for(int j = 0; j < board[i].size(); j++)
            {
                if(board[i][j] != '.' && s.find(board[i][j]) != s.end()) return false;
                s.insert(board[i][j]);
            }
        }
        
        
        for(int j = 0; j < board[0].size(); j++)
        {
            set<char> s;
            for(int i = 0; i < board.size(); i++)
            {
                if(board[i][j] != '.' && s.find(board[i][j]) != s.end()) return false;
                s.insert(board[i][j]);
            }
        }
        
        
        for(int i = 0; i < board.size(); i+=3)
        {
            for(int j = 0; j < board[i].size(); j+=3)
            {
                set<char> s;
                for(int ki = 0; ki < 3; ki++)
                {
                    for(int kj = 0; kj < 3; kj++)
                    {
                        if(board[i + ki][j + kj] != '.' && s.find(board[i + ki][j + kj]) != s.end()) return false;
                        s.insert(board[i + ki][j + kj]);
                    }
                }
            }
        }
        
        return true;
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = {}
        column = {}
        cell = {}
        
        
        m, n = len(board), len(board[0])
        for i in range(m):
            for j in range(n):
                num = board[i][j]
                if num == '.':
                    continue
                
                if i not in row:
                    row[i] = []
                    
                if j not in column:
                    column[j] = []
                
                
                idx = (i//3)*3+(j//3)
                
                if idx not in cell:
                    cell[idx] = []
                
                if num in row[i] or num in column[j] or num in cell[idx]:
                    # print(i, j)
                    # print(num, row, column, cell)
                    return False
                
                row[i].append(num)
                column[j].append(num)
                cell[idx].append(num)
        
        return True